### PR TITLE
Make full-rescan stats publish atomically and skip existing incremental file stats

### PR DESCRIPTION
**Pull request by mcusack-floe**
## Summary

This changes file-group stats persistence so incremental reconciles no longer fail on already-present file stats. Incremental file-group submissions now use putTargetStatsIfAbsent, skip existing records, and only invalidate/cache-count records that were newly created.

For full rescans, file-group chunks no longer overwrite the active stats generation as they complete. Instead, each chunk stages file stats into an unpublished full-rescan-<parentJobId> generation. Snapshot finalization publishes that generation atomically with the merged aggregate stats, so readers continue seeing the previous published stats until finalize succeeds.
Also updates the test override stats store and coverage for staged generation publication, full-rescan finalization, duplicate stats handling, and incremental skip behavior.

## Type of Change

- [ ] feat (new functionality)
- [X] fix (bug fix)
- [ ] docs (documentation only)
- [ ] refactor (no behavior change)
- [ ] test (adds/updates tests)
- [ ] chore (build/tooling/maintenance)

## Testing

Describe how this was validated.

- [X] `make fmt`
- [X] `make verify`
- [X] Added/updated tests for changed behavior

## Deployment and Compatibility

- [X] No breaking changes
- [ ] Breaking changes (describe below)
- [ ] Config/env var changes (describe below)

## Checklist

- [X] PR is scoped and focused
- [ ] Commit messages follow conventional commit style where practical
- [ ] Documentation updated where needed
- [ ] No credentials, tokens, or secrets are included
- [ ] This PR does not disclose a security vulnerability publicly (see `SECURITY.md`)

### CODE REVIEW

File: service/src/main/java/ai/floedb/floecat/service/reconciler/jobs/ReconcilePlannerScheduler.java [hunk 1]
```diff
@@ -57,7 +57,10 @@ public class ReconcilePlannerScheduler {
           java.util.Set.of(
               ReconcileCapturePolicy.Output.TABLE_STATS,
               ReconcileCapturePolicy.Output.FILE_STATS,
-              ReconcileCapturePolicy.Output.COLUMN_STATS));
+              ReconcileCapturePolicy.Output.COLUMN_STATS,
+              ReconcileCapturePolicy.Output.PARQUET_PAGE_INDEX),
+          ReconcileCapturePolicy.DefaultColumnScope.ALL,
```
Comment (mrouvroy-floe) [inline service/src/main/java/ai/floedb/floecat/service/reconciler/jobs/ReconcilePlannerScheduler.java]: [major] **Undocumented capture-policy change bundled into a stats-atomicity fix**

The `DEFAULT_CAPTURE_POLICY` change adds `ReconcileCapturePolicy.Output.PARQUET_PAGE_INDEX` to the outputs and switches the default column scope from `FIRST_N` to `DefaultColumnScope.ALL`. Neither change is mentioned in the PR description, which is scoped to fixing stats persistence.

Every auto-scheduled reconcile will now capture parquet page indexes and column stats for *all* columns rather than the first 32 — a real increase in reconcile work, storage, and cost for every table the scheduler touches.

Split this into its own PR with an explicit justification, or document the intent and expected cost impact here if the change is intentional.

---

🤖 _Generated with Claude Opus 4.8 and GPT-5.5._

File: service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java [hunk 2]
```diff
@@ -504,12 +518,56 @@ private void persistTargetStats(
       return;
     }
     long batchStartNanos = System.nanoTime();
-    statsStore.putTargetStatsBatch(tableId, snapshotId, nonNullStats);
-    statsOrchestrator.invalidateStatsCache(tableId, snapshotId, nonNullStats);
+    List<TargetStatsRecord> created = new java.util.ArrayList<>(nonNullStats.size());
+    for (TargetStatsRecord record : nonNullStats) {
```
Comment (mrouvroy-floe) [inline service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java]: [major] **Incremental stats persist regresses from a batched write to sequential per-record round-trips**

The incremental path previously called `statsStore.putTargetStatsBatch(...)`, which wrote blobs and reserved pointers via `compareAndSetBatch` in batches of up to `MAX_POINTER_BATCH_SIZE` (100). The replacement loop calls `putTargetStatsIfAbsent` once per record.

Each `putTargetStatsIfAbsent` call synchronously performs `ensureActiveGeneration` (pointer read + manifest resolution), `createIfAbsent` (pointer get + blob head + blob put + single-key CAS), and `updateTargetLatestSnapshotIfNewer` (its own CAS-retry loop). For a chunk carrying hundreds of file stats, this means hundreds of sequential DynamoDB/S3 round-trips where there used to be a single batched CAS — materially inflating incremental-reconcile latency on the hot path this PR is targeting.

Consider reintroducing a batch-write abstraction for the incremental path, or confirm that the skip-if-present semantics are worth the latency cost and document it explicitly.

---

🤖 _Generated with Claude Opus 4.8 and GPT-5.5._

File: service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java [hunk 3]
```diff
@@ -504,12 +518,56 @@ private void persistTargetStats(
...
+      if (statsStore.putTargetStatsIfAbsent(record)) {
```
Comment (mrouvroy-floe) [inline service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java]: [major] **Incremental path drops table/snapshot validation, allowing stats pollution across keyspaces**

The previous `putTargetStatsBatch(tableId, snapshotId, nonNullStats)` call passed through `StatsRepository.requireRecordForSnapshot(...)`, which enforced that every record belongs to the supplied `tableId` and `snapshotId`. The new `putTargetStatsIfAbsent(record)` derives the storage key entirely from the record itself with no such check.

A malformed or misdirected worker response can therefore write stats under a different table or snapshot, polluting that keyspace, while the cache invalidation still targets only the planned `tableId`/`snapshotId` — leaving the affected entry stale and the corrupted entry untouched.

Validate each file-stat record against the planned `tableId` and `snapshotId` before entering the loop, or add a table/snapshot assertion inside `putTargetStatsIfAbsent`.

---

🤖 _Generated with Claude Opus 4.8 and GPT-5.5._

File: service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java [hunk 4]
```diff
@@ -504,12 +518,56 @@ private void persistTargetStats(
...
-    metrics.statsItemCount += nonNullStats.size();
+  }
+
+  private void persistDraftFileGroupStats(
+      ReconcileJobStore.LeasedJob lease,
+      ResourceId tableId,
+      ReconcileFileGroupTask plannedTask,
+      List<TargetStatsRecord> fileStats) {
+    long snapshotId = plannedTask.snapshotId();
+    List<TargetStatsRecord> nonNullStats = validateFileStats(tableId, snapshotId, fileStats);
```
Comment (mrouvroy-floe) [inline service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java:540]: [major] **Full-rescan draft generation stages out-of-group file stats that survive into the published manifest**

`persistDraftFileGroupStats` stages every validated file stat into the draft generation, but validation only checks table/snapshot identity — it does not verify that `record.getFile().getFilePath()` is one of the files in `plannedTask.filePaths()`.

The previous full-rescan path filtered the replacement set through `coverage.expectedFiles()`, so extra stats submitted by a buggy worker for files outside its assigned group were silently discarded. With the draft generation path, those records remain in the generation and become active when `publishStatsGeneration` flips the manifest pointer, potentially exposing corrupt or stale stats to readers.

Build a `Set` from `plannedTask.filePaths()` and reject any file-stat record whose canonical file path is not a member before staging it into the draft generation.

---

🤖 _Generated with Claude Opus 4.8 and GPT-5.5._

File: service/src/main/java/ai/floedb/floecat/service/repo/impl/StatsRepository.java [hunk 5]
```diff
@@ -171,6 +171,74 @@ public void putTargetStatsBatch(
     }
   }
 
+  @Override
+  public void replaceTargetStatsInGeneration(
```
Comment (mrouvroy-floe) [inline service/src/main/java/ai/floedb/floecat/service/repo/impl/StatsRepository.java:185]: [minor] **Abandoned full-rescan draft generations are never reclaimed by GC**

`replaceTargetStatsInGeneration` writes record pointers and blobs under the `full-rescan-<parentJobId>` generation but never writes the generation manifest; that only happens in `publishStatsGeneration` at finalize. `deleteUnreferencedGenerations` keys reclamation on manifest existence — a missing manifest causes an unconditional `continue`, deliberately protecting in-flight replaces.

A full rescan that stages some chunks and then never finalizes (job crash, cancellation, or a retry that receives a new plan job and therefore a new `parentJobId`) leaves its staged record pointers and blobs permanently orphaned. GC will never reclaim them, and only a full `deleteAllStatsForSnapshot` prefix wipe removes them — an unbounded leak over time.

Consider a cleanup mechanism: a time-bounded TTL on unfinalized draft generation prefixes, a tombstone written on job failure/cancellation, or an audit pass that removes draft generations older than the max expected job duration.

---

🤖 _Generated with Claude Opus 4.8 and GPT-5.5._

File: service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java [hunk 6]
```diff
@@ -504,12 +518,56 @@ private void persistTargetStats(
...
+        throw new IllegalArgumentException("file-group stats do not match file-group task");
+      }
+    }
+    return nonNullStats;
+  }
+
+  static String statsGenerationId(ReconcileJobStore.LeasedJob lease) {
+    String parentJobId = lease == null || lease.parentJobId == null ? "" : lease.parentJobId.trim();
+    String jobId = lease == null || lease.jobId == null ? "" : lease.jobId.trim();
+    String source = parentJobId.isBlank() ? jobId : parentJobId;
```
Comment (mrouvroy-floe) [inline service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java]: [nit] **`statsGenerationId` fallback to `jobId` risks a silent generation mismatch**

When `parentJobId` is blank, `statsGenerationId` silently falls back to the lease's own `jobId`. Today this is safe because file-group children and the finalize job always carry a non-blank `parentJobId`, so both resolve to the same `full-rescan-<parentJobId>`.

If that invariant broke for one side, file-group chunks would stage into `full-rescan-<childJobId>` while the finalizer would publish `full-rescan-<finalizerJobId>` — a different, empty generation — silently dropping every staged file stat with no error.

Prefer failing loudly (throw on blank `parentJobId`) over a silent fallback, so a mismatch is caught immediately rather than surfacing as mysteriously missing stats.

---

🤖 _Generated with Claude Opus 4.8 and GPT-5.5._

File: service/src/main/java/ai/floedb/floecat/service/reconciler/jobs/ReconcilePlannerScheduler.java (hunk 1 above)
Comment (mcusack-floe) [reply service/src/main/java/ai/floedb/floecat/service/reconciler/jobs/ReconcilePlannerScheduler.java]: Addressed in the latest commit by removing the unrelated default capture-policy change from this PR. `DEFAULT_CAPTURE_POLICY` is back to the existing default outputs (`TABLE_STATS`, `FILE_STATS`, `COLUMN_STATS`) and no longer adds `PARQUET_PAGE_INDEX` or changes the default column scope, so this stats atomicity/GC patch does not alter auto-scheduled capture breadth.

File: service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java (hunk 2 above)
Comment (mcusack-floe) [reply service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java]: Addressed in the latest commit by restoring a batched API with the skip-if-present semantics this path needs. `StatsStore.putTargetStatsBatchIfAbsent(tableId, snapshotId, records)` now lets the incremental path reserve/create the missing records as a batch, and `LeasedFileGroupExecutionService` invalidates cache only for the records actually created. That removes the sequential per-record store loop while preserving idempotent duplicate handling.

File: service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java (hunk 3 above)
Comment (mcusack-floe) [reply service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java]: Addressed in the same batch-if-absent change. The new API keeps `tableId` and `snapshotId` in the method signature, and the repository validates every canonicalized record against that table/snapshot before writing. The latest commit also adds a regression test for rejecting records outside the requested snapshot/keyspace.

File: service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java (hunk 4 above)
Comment (mcusack-floe) [reply service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java:540]: Addressed in the latest commit. Full-rescan draft stats now go through `validateFileStats(...)` for table/snapshot/file-target validation and then check each staged file stat against the planned file-group paths before writing into the draft generation. Out-of-group file stats now fail before staging, so they cannot survive into the published manifest.

File: service/src/main/java/ai/floedb/floecat/service/repo/impl/StatsRepository.java (hunk 5 above)
Comment (mcusack-floe) [reply service/src/main/java/ai/floedb/floecat/service/repo/impl/StatsRepository.java:185]: Addressed in the latest commit with job-aware durable cleanup rather than a time-based guess. Terminal failed/cancelled full-rescan `PLAN_SNAPSHOT` jobs now persist `statsCleanupState=PENDING`, only become eligible after descendants are terminal/finalized, and are revisited by cancellation/projection maintenance until cleanup completes.

The repository cleanup path is also fenced now: publication moves `WRITING -> PUBLISHING -> PUBLISHED`, while cleanup can only claim `WRITING -> DELETING`. `deleteUnpublishedStatsGeneration(...)` returns `DELETED`, `NOT_DELETABLE_PUBLISHED`, or `RETRYABLE_IN_PROGRESS`; `DurableReconcileJobStore` marks cleanup complete only for the first two and leaves `PUBLISHING`/in-progress cases pending for a later maintenance retry.

Regression coverage now includes transient cleanup failure followed by retry, no cleanup while descendants are live, published-generation refusal, the publish/delete interleaving after the final manifest check, and `RETRYABLE_IN_PROGRESS` followed by a later successful delete.

File: service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java (hunk 6 above)
Comment (mcusack-floe) [reply service/src/main/java/ai/floedb/floecat/service/reconciler/impl/LeasedFileGroupExecutionService.java]: Addressed in the latest commit by removing the silent fallback. `statsGenerationId(...)` now requires a nonblank parent job id and throws if that invariant is missing, so a file-group/finalizer mismatch fails loudly instead of writing to a generation derived from the child job id. There is also a regression test for the missing-parent case.

**Review by mrouvroy-floe — APPROVED** [review]
